docs/man/tcmodinfo.1
.TH tcmodinfo 1 "21th January 2003" "tcmodinfo(1)"
.SH NAME
tcmodinfo \- Get and Set information in transcode modules
.SH SYNOPSIS
.na
.B tcmodinfo
[
.B -i
.I name
] [
.B -m
.I path
] [
.B -t
.I type
] [
.B -M
.I param
] [
.B -s
.I socket
] [
.B -p
] [
.B -d
.I verbosity
] [
.B -v
]
.SH COPYRIGHT
\fBtcmodinfo\fP is Copyright (C) by Tilmann Bitterberg/Transcode Team
.SH DESCRIPTION
.B tcmodinfo
loads a supplied transcode(1) module and prints its parameters, or
inquiry about default settings. It can also connect to transcode through
a socket to do runtime configuration of transcode and its filters.
Anyway, this last feature may be moved to some other helper program
in future releases.
.SH OPTIONS
.TP
\fB-i\fP \fIname\fP
Specify the name of the module. \fIName\fP is specified without the prefix
.B (i.e., filter_)_
and without the suffix
.B \.so
So if you want information about filter_smooth.so, just supply
.B smooth
.TP
\fB-m\fP \fIpath\fP
Look in \fIPath\fP instead of the compiled-in module path for a module.
.TP
\fB-t\fP \fItype\fP
Specify the \fIType\fP of module to load. Default is filter modules.
In current (1.1.0) release, tcmodinfo support 'filter', 'encode' and 'multiplex'
module types. More types will be added in future releases, to cover all
the transcode module range.
.TP
\fB-M\fP \fIparameter\fP
Print out the default setting of given parameter of a given (use -i/-t options
too) module. If parameter isn't known, nothing will be print out.
.TP
\fB-s\fP \fIsocket\fP
Connect to socket \fIsocket\fP. You can use this socket to do runtime
configuration of transcode. For all the boring details about the socket
protocol have a look at /docs/filter\-socket.txt. transcode has to be started
with the \fB\-\-socket\fP option to open the socket.
.TP
\fB-d\fP \fIverbosity\fP
Specify the verbosiness level to use, like transcode does. Default value
is 1 (TC_INFO verbosiness).
.TP
.B -p
Print the compiled-in module path and exit.
.TP
.B -v
Print version information and exit.
.SH EXAMPLES
The command
.PP
.B tcmodinfo \-i smooth
.PP
prints information about the smooth filter.
.PP
.B tcmodinfo \-i levels -t filter
.PP
same as above for levels filter (-t filter it's the default)
.PP
.B tcmodinfo \-i raw -t multiplex
.PP
same as above for raw multiplexor
.PP
.B tcmodinfo \-i xvid -t encode
.PP
same as above for XviD encoder
.PP
.B transcode \-\-socket /tmp/tc\-socket &
.br
.B echo help | tcmodinfo \-s /tmp/tc\-socket
.PP
prints the socket help text and exits.
.PP
.B tcmodinfo \-s /tmp/tc\-socket
.PP
Will drop you into an interactive "shell" where you can talk to the socket.
.SH AUTHORS
.B tcmodinfo
was written by Tilmann Bitterberg
.br
<transcode@tibit.org> with contributions from
many others.  See AUTHORS for details.
.SH SEE ALSO
.BR avimerge (1),
.BR avisplit (1),
.BR tccat (1),
.BR tcdecode (1),
.BR tcdemux (1),
.BR tcextract (1),
.BR tcprobe (1),
.BR tcscan (1),
.BR transcode (1)



ChangeLog
2006-11-15 Francesco Romani <fromani at gmail dot com>
  * src/cmdline_def.h docs/man/transcode.1
  review a bit --no_log_color option and update manpage accordingly.

2006-11-12 Francesco Romani <fromani at gmail dot com>
  * Added files: tools/tcstub.c tools/tcstub.h
  refactorize and extract stub code needed by helper tools.

  * tools/Makefile.am tools/tcmodinfo.c tools/tcexport.c
  port to new stub code.

  * tools/Makefile.am tools/aud_scan.c tools/aud_scan_avi.c
    tools/avifix.c tools/aviindex.c tools/avimerge.c tools/avimisc.c
    tools/avisplit.c tools/avisync.c tools/buffer.c tools/rawsource.c
    tools/tcmp3cut.c
  add format helper trailer.

  * docs/man/tcmodinfo.1
  minor fixes.

  * added file: tools/tcmodchain.c

  * tools/Makefile.am
  new helper, still experimental, tool to let the user check the
  compatibility between some modules before to load transcode.
  Manpage will be added soon.

2006-11-11 Francesco Romani <fromani at gmail dot com>
  * Added file: libtc/tcformat.h
  Add known formats specifier much like already done for codecs.

  * libtc/tcmodule-info.h
  Add supported formats for modules, useful for de/multiplexors.

  * libtc/libtc.h libtc/tccodecs.c libtc/tcmoduleinfo.c
  Since we have now TCCodecID, use it instead of plain `int'.

  * libtc/tcmodule.c
  fill format fields for dummy internal class too.

  * encode/encode_copy.c encode/encode_faac.c encode/encode_lame.c
    encode/encode_lzo.c encode/encode_null.c encode/encode_x264.c
    encode/encode_xvid.c export/export_pvn.c filter/filter_doublefps.c
    filter/filter_levels.c filter/filter_null.c import/import_pv3.c
    import/import_pvn.c import/import_vag.c import/import_x11.c
    import/nuv/import_nuv.c multiplex/multiplex_avi.c
    multiplex/multiplex_null.c multiplex/multiplex_raw.c
    multiplex/multiplex_y4m.c testsuite/test-tcmoduleinfo.c
  switch to TCCodecID/TCFormatID when declaring module properties.
  (In filling TCModuleInfo)

  * encode/encode_lzo.c
  staticize encoder buffer as part as module's PrivateData.

2006-11-11 Andrew Church <achurch at achurch dot org>
  * import/import_pvn.c
  Fix processing of floating-point files with negative maxvals.

  * export/export_pvn.c
  Write grayscale (PV5a) files when -K is given to transcode.

2006-11-09 Andrew Church <achurch at achurch dot org>
  * configure.in
  Add AM_PROG_CC_C_O as requested by automake 1.10.

  * import/import_pvn.c
  Add workaround for broken compilers that can't align stack-based structures
  (and update/clean up the #if's).

2006-11-05 Francesco Romani <fromani at gmail dot com>
  * libtc/tcmodule-plugin.h
  use parameter name in log.

  * libtc/tccodecs.h
  from #define to enum.

2006-11-01 Francesco Romani <fromani at gmail dot com>
  * encode/encode_lame.c encode/encode_lzo.c encode/encode_x264.c
  switch to libtc macros whenever possible (and reasonnable).

  * multiplex/multiplex_avi.c
  provide default fallback fourcc string.

  * import/tcprobe.c
  fix output glitch.

  * encode/encode_lame.c
  properly handle MONO audio.

2006-10-23 Francesco Romani <fromani at gmail dot com>
  * src/framebuffer.c
  enlarge audio buffers too (see new comment) to make
  encoded buffers suitable for encoder flush and make lame happy.
  For post-1.1.0 buffer (size) management really need to be reviewd
  again.

  * src/encoder.c libtc/tcmodule-data.h encode/encode_copy.c
    encode/encode_lame.c
  Introduce flushing support for audio encoder.
  Currently BROKEN (need more testing, and encoder core deserve
  a refactoring and more documentation).

2006-10-23 Francesco Romani <fromani at gmail dot com>
  * import/decode_lavc.c import/extract_avi.c src/probe.c
  TC_CODEC_MPG1 must die (too generic, redundant).

  * encode/encode_copy.c encode/encode_faac.c encode/encode_lame.c
    encode/encode_lzo.c encode/encode_null.c encode/encode_x264.c
    encode/encode_xvid.c
  move return codes to new TC_OK/TC_ERROR

2006-10-21 Andrew Church <achurch at achurch dot org>
  * import/import_vag.c
  Trivial debug message fix.

2006-10-20 Francesco Romani <fromani at gmail dot com>
  * libtc/libtc.h
  add TC_{OK,ERROR,UNKNOWN} generic enums, so we can
  start to get rid of TC_IMPORT/EXPORT_* counterparts.
  New enums use back-compatible values.

2006-10-19 Francesco Romani <fromani at gmail dot com>
  * src/transcode.c src/transcode.h src/socket.c import/import_dvd.c
  remove vob_t.in_flag field since nobody uses it anymore
  (checked with plain old grep -R).

  * import/tcprobe.c
  implement first draft of new output mode.

2006-10-18 Francesco Romani <fromani at gmail dot com>
  * import/tcprobe.c
  restore old output (well, almost identical :) ) as default
  behaviour for tcprobe, and add a new option to toggle new
  fancy output (not yet implemented), much like raw output mode.

2006-10-17 Andrew Church <achurch at achurch dot org>
  * import/import_pvn.c
  Allow a framerate field of 0 (defaulting to 15 fps), and do not use SSE2
  code if SSE2 acceleration is disabled (--accel) or unavailable.

2006-10-15 Andrew Church <achurch at achurch dot org>
  * acinclude.m4
  Modify TC_PKG_* macros to use variables rather than temporary files
  for package reports.

  * configure.in
  Update another obsolete macro.

2006-10-14 Andrew Church <achurch at achurch dot org>
  * encode/encode_x264.c
  x264 API change: i_rf_constant -> f_rf_constant

  * configure.in
  Check x264 API version; also update obsolete autoconf macros.

2006-10-14 Andrew Church <achurch at achurch dot org>
  * import/import_pvn.c
  Fix divide-by-zero bug with negative maxvals.
  Handle reading from pipes properly.
  Allow "-" for input from stdin.

  * export/export_pvn.c
  Fix typo in module name.
  Handle writing to pipes properly.
  Allow "-" for output to stdout.

  * src/cmdline.c
  Allow "-" as an input filename.

  * src/cmdline_def.h
  Allow "-" as an output filename.

  * src/cmdline.c src/counter.c src/decoder.c src/dl_loader.c src/encoder.c

tools/tcmodinfo.c
/*
 *  tcmodinfo.c
 *
 *  Copyright (C) Tilmann Bitterberg - August 2002
 *  updated and partially rewritten by
 *  Copyright (C) Francesco Romani - January 2006
 *
 *  This file is part of transcode, a video stream processing tool
 *
 *  transcode is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2, or (at your option)
 *  any later version.
 *
 *  transcode is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with GNU Make; see the file COPYING.  If not, write to
 *  the Free Software Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.
 *
 */

#include "tcstub.h"

#define EXE "tcmodinfo"

enum {
    STATUS_OK = 0,
    STATUS_BAD_PARAM,
    STATUS_NO_MODULE,
    STATUS_MODULE_ERROR,
    STATUS_NO_SOCKET,
    STATUS_SOCKET_ERROR,
    STATUS_BAD_MODULES,
};

void version(void)
{
    printf("%s (%s v%s) (C) 2001-2006 Tilmann Bitterberg, "
           "transcode team\n", EXE, PACKAGE, VERSION);
}

static void usage(void)
{
    version();
    tc_log_info(EXE, "Usage: %s [options]", EXE);
    fprintf(stderr, "    -i name           Module name information (like \'smooth\')\n");
    fprintf(stderr, "    -p                Print the compiled-in module path\n");
    fprintf(stderr, "    -d verbosity      Verbosity mode [1 == TC_INFO]\n");
    fprintf(stderr, "    -m path           Use PATH as module path\n");
    fprintf(stderr, "    -M element        Request to module informations about <element>\n");
    fprintf(stderr, "    -s socket         Connect to transcode socket\n");
    fprintf(stderr, "    -t type           Type of module (filter, encode, multiplex)\n");
    fprintf(stderr, "\n");
}

static int do_connect_socket(const char *socketfile)
{
    int sock, retval;
    struct sockaddr_un server;
    char buf[OPTS_SIZE];
    fd_set rfds;
    struct timeval tv;
    ssize_t n;

    sock = socket(AF_UNIX, SOCK_STREAM, 0);
    if (sock < 0) {
        perror("opening stream socket");
        return STATUS_NO_SOCKET;
    }
    server.sun_family = AF_UNIX;
    strlcpy(server.sun_path, socketfile, sizeof(server.sun_path));

    if (connect(sock, (struct sockaddr *) &server,
                sizeof(struct sockaddr_un)) < 0) {
        close(sock);
        perror("connecting stream socket");
        return STATUS_NO_SOCKET;
    }

    while (1) {
    /* Watch stdin (fd 0) to see when it has input. */
    FD_ZERO(&rfds);
    FD_SET(0, &rfds); // stdin
    FD_SET(sock, &rfds);
    /* Wait up to five seconds. */
    tv.tv_sec = 5;
    tv.tv_usec = 0;

    retval = select(sock+1, &rfds, NULL, NULL, NULL);
    /* Don't rely on the value of tv now! */

    memset(buf, 0, sizeof (buf));  // null-termination in advance, slowly

    if (retval>0) {
        if (FD_ISSET(0, &rfds)) {
        fgets(buf, OPTS_SIZE, stdin);
        }
        if (FD_ISSET(sock, &rfds)) {
        if ( (n = read(sock, buf, OPTS_SIZE)) < 0) {
            perror("reading on stream socket");
            break;
        } else if (n == 0) { // EOF
            fprintf (stderr, "Server closed connection\n");
            break;
        }
        printf("%s", buf);
        continue;
        }
    }

    if (write(sock, buf, strlen(buf)) < 0)
        perror("writing on stream socket");

    memset(buf, 0, sizeof (buf));

    if (read(sock, buf, OPTS_SIZE) < 0)
        perror("reading on stream socket");

    printf("%s", buf);

    if (!isatty(0))
        break;
    }

    close(sock);
    return STATUS_OK;
}

int main(int argc, char *argv[])
{
    int ch;
    const char *filename = NULL;
    const char *modpath = MOD_PATH;
    const char *modtype = "filter";
    const char *modarg = NULL;
    const char *socketfile = NULL;
    char options[OPTS_SIZE] = { '\0', };
    int print_mod = 0;
    int connect_socket = 0;
    int ret = 0;
    int status = STATUS_NO_MODULE;

    /* needed by filter modules */
    TCVHandle tcv_handle = tcv_init();
    TCFactory factory = NULL;
    TCModule module = NULL;

    vframe_list_t ptr;

    memset(&ptr, 0, sizeof(ptr));

    ac_init(AC_ALL);
    tc_set_config_dir(NULL);

    if (argc == 1) {
        usage();
        return STATUS_BAD_PARAM;
    }

    while(1) {
        ch = getopt(argc, argv, "d:i:?vhpm:M:s:t:");
        if (ch == -1) {
            break;
        }

        switch (ch) {
          case 'd':
            if (optarg[0] == '-') {
                usage();
                return STATUS_BAD_PARAM;
            }
            verbose = atoi(optarg);
            break;
          case 'i':
            if (optarg[0] == '-') {
                usage();
                return STATUS_BAD_PARAM;
            }
            filename = optarg;
            break;
          case 'm':
            modpath = optarg;
            break;
          case 'M':
            modarg = optarg;
            break;
          case 't':
            if (!optarg) {
                usage();
                return STATUS_BAD_PARAM;
            }

            if (!strcmp(optarg, "filter")
             || !strcmp(optarg, "encode")
             || !strcmp(optarg, "multiplex")) {
                modtype = optarg;
            } else {
                modtype = NULL;
            }
            break;
          case 's':
            if (optarg[0] == '-') {
                usage();
                return STATUS_BAD_PARAM;
            }
            connect_socket = 1;
            socketfile = optarg;
            break;
          case 'p':
            print_mod = 1;
            break;
          case 'v':
            version();
            return STATUS_OK;
          case '?': /* fallthrough */
          case 'h': /* fallthrough */
          default:
            usage();
            return STATUS_OK;
        }
    }

    if (print_mod) {
        printf("%s\n", modpath);
        return STATUS_OK;
    }

    if (connect_socket) {
        do_connect_socket(socketfile);
        return STATUS_OK;
    }

    if (!modtype || !strcmp(modtype, "import")) {
        tc_log_error(EXE, "Unknown module type (not in filter, encode, multiplex)");
        return STATUS_BAD_PARAM;
    }

    if (!filename) {
        usage();
        return STATUS_BAD_PARAM;
    }

    /* 
     * we can't distinguish from OMS and NMS modules at glance, so try
     * first using new module system
     */
    factory = tc_new_module_factory(modpath, verbose);
    module = tc_new_module(factory, modtype, filename);

    if (module != NULL) {
        const char *answer = NULL;

        if (verbose >= TC_DEBUG) {
            tc_log_info(EXE, "using new module system");
        }
        if (verbose >= TC_INFO) {
            /* overview and options */
            tc_module_inspect(module, "help", &answer);
            puts(answer);
            /* module capabilities */
            tc_module_show_info(module, verbose);
            /* current configuration */
            puts("\ndefault module configuration:");
            tc_module_inspect(module, "all", &answer);
            puts(answer);
        }
        if (modarg != NULL) {
            tc_log_info(EXE, "informations about '%s' for "
                             "module:", modarg);
            tc_module_inspect(module, modarg, &answer);
            puts(answer);
        }
        tc_del_module(factory, module);
        status = STATUS_OK;
    } else if (!strcmp(modtype, "filter")) {
        char namebuf[NAME_LEN];
        /* compatibility support only for filters */
        if (verbose >= TC_DEBUG) {
            tc_log_info(EXE, "using old module system");
        }
        /* ok, fallback to old module system */
        filter[0].name = namebuf;
        tc_snprintf(filter[0].name, NAME_LEN, "%s", filename);
    
        ret = load_plugin(modpath, 0, verbose);
        if (ret == 0) {
            strlcpy(options, "help", OPTS_SIZE);
            ptr.tag = TC_FILTER_INIT;
            if ((ret = filter[0].entry(&ptr, options)) != 0) {
                status = STATUS_MODULE_ERROR;
            } else {
                memset(options, 0, OPTS_SIZE);
                ptr.tag = TC_FILTER_GET_CONFIG;
                ret = filter[0].entry(&ptr, options);

                if (ret == 0) {
                    if (verbose >= TC_INFO) {
                        fputs(options, stdout);
                    }
                    status = STATUS_OK;
                }
            }
        }
   }

   ret = tc_del_module_factory(factory);
   tcv_free(tcv_handle);
   return status;
}

/*************************************************************************/

/*
 * Local variables:
 *   c-file-style: "stroustrup"
 *   c-file-offsets: ((case-label . *) (statement-case-intro . *))
 *   indent-tabs-mode: nil
 * End:
 *
 * vim: expandtab shiftwidth=4:
 */
